### PR TITLE
4467: Adgangsplatformen single signout

### PR DESCRIPTION
**Pull request by kasperg**
#### Link to issue

https://platform.dandigbib.org/issues/4467

#### Description

This PR updates the Ding Adgangsplatformen module to support and use [single signout](https://github.com/DBCDK/hejmdal/blob/master/docs/single-logout.md).

A new endpoint is defined which should live up to the provided specification. The module will return status code 403 for users which should not support single signout because they have not authenticated via Adgangsplatformen. @vibjerg has nodded to this approach although it is not mentioned in the current documentation.

#### Checklist

- [x] My complies with [our coding guidelines](../docs/code_guidelines.md).
- [x] My code passes our static analysis suite. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
- [x] My code passes our continuous integration process. If not then I have added a comment explaining why this change should be exempt from the code standards and process.

### CODE REVIEW

File: modules/ding_adgangsplatformen/ding_adgangsplatformen.module [hunk 1]
```diff
@@ -209,7 +216,16 @@ function ding_adgangsplatformen_logout($hooked = FALSE) {
   $token = $_SESSION['oauth2token'];
 
   // Generate logout request for the authorization service and send the request.
-  $logout_url = $config['urlLogout'] . '?access_token=' . $token . '&redirect_uri=' . $base_url;
+  $logout_url = url(
+    $config['urlLogout'],
+    array(
+      'external' => true,
+      'query' => array(
+        'singlelogout' => 'true',
```
Comment (cableman) [inline modules/ding_adgangsplatformen/ding_adgangsplatformen.module]: This parameter should be configurable to allow to opt-out of singe sign out.

Comment (kasperg) [reply modules/ding_adgangsplatformen/ding_adgangsplatformen.module]: :+1: Done.

File: modules/ding_adgangsplatformen/ding_adgangsplatformen.module [hunk 2]
```diff
@@ -225,6 +243,41 @@ function ding_adgangsplatformen_logout($hooked = FALSE) {
...
+    session_destroy();
+  }
+
+  // Logout requests must never be cached. A cached response would cause
+  // Adgangsplatformen to think the user was logged out when it fact he was not.
+  drupal_page_is_cacheable(FALSE);
+  drupal_add_http_header('Status', "$status_code $reason");
+  drupal_add_http_header(
+    'X-Frame-Options',
+    'allow-from https://login.bib.dk'
```
Comment (cableman) [inline modules/ding_adgangsplatformen/ding_adgangsplatformen.module]: Maybe this should be put into configuration as an hidden variable as this endpoint have changed once before.

Comment (kasperg) [reply modules/ding_adgangsplatformen/ding_adgangsplatformen.module]: @cableman: I see several options here.

1. Add a hidden variable
2. Use the hostnames of one of the urls from the OAuth provider (assuming that the iframe will be located on the same system.
3. A combination of 1 and 2 where the hidden variable falls back back on the hostname.

Comment (cableman) [reply modules/ding_adgangsplatformen/ding_adgangsplatformen.module]: @kasperg I was thinking something like https://github.com/ding2/ding2/blob/master/modules/ding_adgangsplatformen/includes/ding_adgangsplatformen.admin.inc#L43 so the solution was the same for these variables. That should not be changed but might need to be in the future...

Comment (kasperg) [reply modules/ding_adgangsplatformen/ding_adgangsplatformen.module]: I have tried to implement this in da9824e.

**Comment by kasperg** [comment]
I think the issues identified by Scrutinizer are out of scope for this PR.

File: modules/ding_provider/connie/connie.auth.inc [hunk 3]
```diff
@@ -7,7 +7,7 @@
 /**
  * Implements hook_auth_single_sign_on().
  */
-function fbs_auth_single_sign_on($name) {
+function connie_auth_single_sign_on($name) {
```
Comment (cableman) [inline modules/ding_provider/connie/connie.auth.inc]: 👍

**Pull request merged**